File: schemas/analytics-metrics/convex/experiment_goals.ts
// experiment_goals: Links experiments to goals.
// See README.md for full design rationale.

import { defineTable } from "convex/server";
import { v } from "convex/values";

export const experiment_goals = defineTable({
  experimentId: v.id("experiments"),
  goalId: v.id("goals"),
  isPrimary: v.boolean(),
})
  .index("by_experiment_id_and_goal_id", ["experimentId", "goalId"])
  .index("by_goal_id", ["goalId"]);

File: schemas/analytics-metrics/spacetimedb/experiment_goals.rs
// experiment_goals: Links experiments to their target conversion goals.
// See README.md for full design rationale.

use spacetimedb::Timestamp;

// Composite unique constraint (not expressible inline):
// - unique(experiment_id, goal_id)

#[spacetimedb::table(name = experiment_goals, public)]
pub struct ExperimentGoal {
    #[primary_key]
    pub id: String, // UUID

    #[index(btree)]
    pub experiment_id: String, // UUID, FK → experiments.id (cascade delete)

    #[index(btree)]
    pub goal_id: String, // UUID, FK → goals.id (cascade delete)

    pub is_primary: bool,
    pub created_at: Timestamp,
}
